### DISCUSSION TITLE
v5.4.2-pre.1

### DISCUSSION THREAD

**Discussion by coolsoftwaretyler** [mobxjs/mobx-state-tree#2160] [General] [answered]
Version 5.4.2-pre.1 is an important build for everyone to test out, because it includes some TypeScript changes that could be seen as either bug fixes or breaking changes.

## RFC - should we consider these changes bug fixes (bump to 5.4.2) or breaking changes (bump to 6.0.0)

```ts
import { t } from "mobx-state-tree";

/**
 * In MobX-State-Tree 5.4.1, this is typed as:
 * ISimpleType<"Red" | "Orange" | "Green">
 */
const namedEnum = t.enumeration("Color", ["Red", "Orange", "Green"]);

/**
 * In MobX-State-Tree 5.4.1, this is typed as:
 * ISimpleType<string>
 */
const anonymousEnum = t.enumeration(["Red", "Orange", "Green"]);

/**
 * If you use mobx-state-tree@5.4.2-pre.1, both of these will be typed as:
 * ISimpleType<"Red" | "Orange" | "Green">
 */
```

[CodeSandbox for version 5.4.1](https://codesandbox.io/p/sandbox/enum-typing-changes-after-5-4-1-k6krfh?file=%2Fsrc%2Findex.ts%3A10%2C15)

[CodeSandbox for version 5.4.2-pre.1](https://codesandbox.io/p/sandbox/enum-typing-changes-after-5-4-1-forked-4sxkkd?file=%2Fsrc%2Findex.ts%3A14%2C1)

It's reasonable to call this change a "bug fix", but for projects that relied on the prior behavior, a patch version might "break" their TypeScript types, if they've typed around our existing bug.

The change comes from #2151, which also "fixes" #1525 and #1664 again, by changing types.

We have also [removed `NonEmptyObject`](https://github.com/mobxjs/mobx-state-tree/pull/2152). If a project had relied on that for any kind of type casting, I think that could also be seen as a breaking change.

And of course, we've moved to TypeScript `5.3.3`, which shouldn't have a direct impact downstream, but we have previously [only called out TS 3.0 or later](https://mobx-state-tree.js.org/tips/typescript). This is not strictly a breaking change, and it's technically in line with "TypeScript 3.0 or later", but it could be seen as disruptive to move so far ahead in TypeScript without ample warning in our version.

## Breaking Changes

* Improved typing for union types by @thegedge in https://github.com/mobxjs/mobx-state-tree/pull/2151 (maybe, see introduction)
* Eliminate `NonEmptyObject` by @thegedge in https://github.com/mobxjs/mobx-state-tree/pull/2152 (maybe, see introduction)
* Bump typescript from 3.9.10 to 5.3.3 by @thegedge in https://github.com/mobxjs/mobx-state-tree/pull/2146

## Features

* No new features

## Fixes

* Some might consider #2146, #2151, and #2152 to be bug fixes, rather than breaking changes.

## Tests

* No test-only additions.

## Docs

* Docs/update docs homepage by @coolsoftwaretyler in https://github.com/mobxjs/mobx-state-tree/pull/2141
* chore: version bump to 5.4.2-pre.1 by @coolsoftwaretyler in https://github.com/mobxjs/mobx-state-tree/pull/2159
* docs: fix typos in docs update by @coolsoftwaretyler in https://github.com/mobxjs/mobx-state-tree/pull/2142

## Community/Developer changes

* Chore/use bun as a package manager and script runner by @coolsoftwaretyler in https://github.com/mobxjs/mobx-state-tree/pull/2148

## New Contributors
* @thegedge made their first contribution in https://github.com/mobxjs/mobx-state-tree/pull/2146

**Full Changelog**: https://github.com/mobxjs/mobx-state-tree/compare/v5.4.1...v5.4.2-pre.1

<hr /><em>This discussion was created from the release <a href='https://github.com/mobxjs/mobx-state-tree/releases/tag/v5.4.2-pre.1'>v5.4.2-pre.1</a>.</em>

**Comment by coolsoftwaretyler** [comment]
Another reason to consider a new major: 

There's another PR that we've left open for a long time which we could finally merge and ship with a new major version: https://github.com/mobxjs/mobx-state-tree/pull/2039.

That PR is much more clearly breaking - it introduces a new scenario in which MST will throw an error. 

It could be nice to move up to `6.0.0` and ship that change, but this still feels like a very small list of breaking changes to make.

**Reply by jamonholmgren** [reply]
I think having this reason to bump major makes it more compelling than just a typescript bugfix.

Anything else in our current issues / PRs that could be released too as small breaking changes?

With that said, I'm also in favor of lightweight majors -- that is, bumping major more often, but having it be a relatively painless thing to upgrade because it doesn't contain much.

**Reply by coolsoftwaretyler** [reply]
I don't know of anything else ready to merge that would be breaking, but if you're happy with more frequent major updates, that resolves some of my own trepidation. I don't want us to feel like we're churning too hard on majors, but I agree this is a good approach.

I will keep this discussion and pre-release available for more discussion, but I'm leaning this direction right now.

**Reply by coolsoftwaretyler** [reply]
I'm not sure if https://github.com/mobxjs/mobx-state-tree/pull/2073 is ready to go yet, and I'm not sure how much work it will take to get it ready. But if I can get it working in the next week or two, it would be another bug-fix-that-breaks, and maybe a good thing to bundle into v6.

**Comment by coolsoftwaretyler** [comment, marked as answer]
Considering [Jamon's opinion](https://github.com/mobxjs/mobx-state-tree/discussions/2160#discussioncomment-8645317), and conversations we had at the last maintainer meeting, I think we're going to bundle these changes into a major version. We'll do a preview release and keep it in preview for a while in case anyone else is sitting on or needs breaking changes we can bundle together.

I'll release that today, and over the next few weeks put together a migration guide.

**Comment by coolsoftwaretyler** [comment]
Release post for the pre-reelease of 6.0.0: https://github.com/mobxjs/mobx-state-tree/releases/tag/v6.0.0-pre.2
